### PR TITLE
[WIP] add checks for obsolete settings

### LINKED ISSUE #1698: Check site config for obsolete settings
Several config settings where removed, renamed, .. from the site configuration in the past.

Could it be useful to `WARN` about obsolete settings and guide to the docs/release notes for migrations hints?

Should that be part of `scripts/check_site.lua`?

### PR DESCRIPTION
**Pull request by bobcanthelpyou**
This might fix #1698 

When running `make V=s GLUON_RELEASE="master" GLUON_TARGET="ar71xx-tiny" GLUON_DEVICES="tp-link-tl-wr840n-v2"` (site conf of Freifunk Hamburg and Gluon master) the following output is printed eight times:
```
site.conf obsolete: opkg.lede is obsolete and could be removed, check the release notes and documentation please
```

I am not sure if `print()` is a good choice or if something like `io.stderr:write()` should be used
e.g. `io.stderr:write(src .. ' obsolete: ' .. string.format(...) .. '\n')` (but it is visible with `V=s`, only)

Also not sure if a generic text is fine or if it should be possible to describe what is wrong, add a link to the documentation/release notes, ... what do you think?

### CODE REVIEW

File: scripts/check_site.lua [hunk 1]
```diff
@@ -108,14 +107,27 @@ local function var_error(path, val, msg)
...
 	local found = 'unset'
 	if val ~= nil then
 		found = string.format('%s (a %s value)', tostring(val), type(val))
 	end
 
-	config_error(src, 'expected %s to %s, but it is %s', path_to_string(path), msg, found)
+	config_error(conf_src(path), 'expected %s to %s, but it is %s', path_to_string(path), msg, found)
 end
 
+function obsolete(path)
```
Comment (mweinelt) [inline scripts/check_site.lua]: allow for custom obsoletion messages to be passed

Comment (bobcanthelpyou) [reply scripts/check_site.lua]: added the possibility to add custom messages

**Comment by mweinelt** [comment]
Being visible only when `V=s` is passed is not good enough.

**Comment by bobcanthelpyou** [comment]
@mweinelt where should it be visible? e.g. an error of the site config is visible via `V=s` only isn't it?

**Comment by NeoRaider** [comment]
One option would be to make obsolete settings a hard failure.

We currently don't have a good way to make warnings visible without `V=s` (or `BUILD_LOGS=1` for logging into files)... but I assume such a facility could be added to OpenWrt.

**Comment by rotanid** [comment]
hard failure should be OK as its easily fixable and clearly highlights something has to be done.

**Comment by bobcanthelpyou** [comment]
running
```
make -j 1 V=s GLUON_BRANCH=${GLUON_BRANCH} GLUON_TARGET=ar71xx-tiny GLUON_RELEASE=${GLUON_RELEASE}
```
with a `probability` setting will fail with
```
*** site.conf error: autoupdater.branches.stable.probability is obsolete Use GLUON_PRIORITY in site.mk instead.
postinst script ./usr/lib/opkg/info/gluon-autoupdater.postinst has failed with exit code 1
```
or with a `opkg.lede` setting will fail with
```
*** site.conf error: opkg.lede is obsolete Use opkg.openwrt instead.
postinst script ./usr/lib/opkg/info/gluon-core.postinst has failed with exit code 1
```
or a different obsolete setting '...' without a custom message will fail with
```
*** site.conf error: ... is obsolete and must be migrated or removed, check the release notes and documentation please
```

**Comment by NeoRaider** [comment]
Implementation looks ok. I think the following formatting/wording would look better:
```
*** site.conf error: ... is obsolete and must be migrated or removed. Please check the release notes and documentation for details.
```

```
*** site.conf error: autoupdater.branches.stable.probability is obsolete. Use GLUON_PRIORITY in site.mk instead.
```

**Comment by bobcanthelpyou** [comment]
@NeoRaider i reworded the message somehow. Not sure if it is okay like it is now: 38a10d5ecec139f7dc70a8120d699b7b8c44c250

Should that be documented somewhere? Are there docs for check_site available?

File: scripts/check_site.lua [hunk 2]
```diff
@@ -315,6 +317,20 @@ function need_domain_name(path)
...
 end
 
+function obsolete(path, msg)
+	local val = loadvar(path)
+	if val == nil then
+		return nil
+	end
+
+	if not msg then
+		msg = 'Please remove or migrate it, check the release notes and documentation for details.'
```
Comment (NeoRaider) [inline scripts/check_site.lua]: Separating these two sentences with a comma looks weird, I'd only use such a construction when the second sentence is very short ("%s is obsolete, please remove or migrate it." was fine IMO). I guess we could just remove the "Please remove or migrate it" completely and make the default msg "Check the release notes and documentation for details.".

Comment (bobcanthelpyou) [reply scripts/check_site.lua]: Removed it. Stashed commit and reword commit msg.

**Pull request merged**